File: 35-search-insert-position/search-insert-position.cpp
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
         int start_index = 0; // snake_case
        int end_index = nums.size() - 1; // snake_case
        
        while (start_index <= end_index) {
            // Integer overflow se bachne ke liye safe formula
            int mid_index = start_index + (end_index - start_index) / 2; // snake_case
            
            // Case 1: Target milne par index return karo
            if (nums[mid_index] == target) {
                return mid_index;
            }
            // Case 2: Agar target bada hai, toh right side jao
            else if (nums[mid_index] < target) {
                start_index = mid_index + 1;
            }
            // Case 3: Agar target chota hai, toh left side jao
            else {
                end_index = mid_index - 1;
            }
        }
        
        // Loop break hone par start_index par hi sahi insertion point hota hai
        return start_index;
    }
};

File: 35-search-insert-position/search-insert-position.java
class Solution {
    public int searchInsert(int[] nums, int target) {
         int startIndex = 0; // camelCase
        int endIndex = nums.length - 1; // camelCase
        
        while (startIndex <= endIndex) {
            // Integer overflow se bachne ke liye safe formula
            int midIndex = startIndex + (endIndex - startIndex) / 2; // camelCase
            
            // Case 1: Agar target mil gaya
            if (nums[midIndex] == target) {
                return midIndex;
            }
            // Case 2: Agar target bada hai, toh right side me dhoondho
            else if (nums[midIndex] < target) {
                startIndex = midIndex + 1;
            }
            // Case 3: Agar target chota hai, toh left side me dhoondho
            else {
                endIndex = midIndex - 1;
            }
        }
        
        // Loop tootne par startIndex hi sahi insertion position (future index) hota hai
        return startIndex;
    }
}